Improving System.Linq.Intersect with no comparer.

varocarbas opened the pull request:
The `Intersect` overload not taking a `IEqualityComparer<TSource> comparer` argument is unnecessarily slow because of calling `IntersectIterator`, whose algorithm is mostly meant to deal with a non-null `comparer`. 

The proposed modification relies on `Contains` and shows a noticeable performance improvement. Below these lines I am including a simplistic testing code where, under my specific conditions, `Intersect2` (i.e., the proposed modification) is systematically 40% quicker.

UPDATE: as rightly pointed out via comments, the original test didn't account for the iterator peculiarities (e.g., being called from a `forearch` loop) where the new approach doesn't perform too well (some times better, even notably better; but not always as shown in the updated code below these lines).

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace TestIntersect
{
    static class Program
    {
        public delegate IEnumerable<TSource> target<TSource>(IEnumerable<TSource> first, IEnumerable<TSource> second);
        public static void Main(string[] args)
        {
            int count = 0;
            int maxCount = 1000;
            int maxItems = 50000;

            List<string> first = new List<string>()
            {
                "abcd", "mdddfg", "bbbbbbb", "999999999", "identical"
            };

            List<string> second = new List<string>()
            {
                "abcd", "123456789", "bbbbbbb", "1255555", "identical"
            };

            while (count < maxItems)
            {
                count++;
                first.Add("13456");
                first.Add("dd");
                first.Add("abceeee");
                first.Add("99999");
                second.Add("93456");
                second.Add("a");
                second.Add("abceeee");
                second.Add("5555");
            };

            target<string>[] targets = new target<string>[] { Intersect, Intersect2 };
            string[] captions = new string[]{ "Original", "New" };
            bool[] iteratorNots = new bool[] { false, true, true, false };

            foreach (bool iteratorNot in iteratorNots)
            {
                for (int i = 0; i < captions.Length; i++)
                {
                    Test(maxCount, targets[i], first, second, captions[i], iteratorNot);
                }
            }

            Console.ReadLine();
        }

        public static void Test(int maxCount, target<string> target, IEnumerable<string> first, IEnumerable<string> second, string caption, bool testIterator)
        {
            Stopwatch sw = new Stopwatch();
            sw.Start();
            if (testIterator) TestIterator(maxCount, target, first, second);
            else TestReturnedCollection(maxCount, target, first, second);
            sw.Stop();

            caption += " " + (testIterator ? "Iterator" : "ReturnedCollection") + " ";

            Console.WriteLine(caption + sw.ElapsedMilliseconds.ToString());
        }

        public static void TestIterator(int maxCount, target<string> target, IEnumerable<string> first, IEnumerable<string> second)
        {
            int count = 0;
            while (count < maxCount)
            {
                count++;
                foreach (var item in target(first, second)) { }
            }
        }

        public static void TestReturnedCollection(int maxCount, target<string> target, IEnumerable<string> first, IEnumerable<string> second)
        {
            int count = 0;
            while (count < maxCount)
            {
                count++;
                target(first, second);
            }
        }

        public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second)
        {
            if (first == null)
            {
                throw new Exception();
            }

            if (second == null)
            {
                throw new Exception();
            }

            return IntersectIterator(first, second, null);
        }

        public static IEnumerable<TSource> Intersect2<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second)
        {
            if (first == null)
            {
                throw new Exception();
            }

            if (second == null)
            {
                throw new Exception();
            }

            IEnumerable<TSource> first2 = first.Distinct();
            IEnumerable<TSource> second2 = second.Distinct();

            foreach (TSource element in first2)
            {
                if (second2.Contains(element))
                {
                    yield return element;
                }
            }
        }

        private static IEnumerable<TSource> IntersectIterator<TSource>(IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
        {
            Set<TSource> set = new Set<TSource>(comparer);
            foreach (TSource element in second)
            {
                set.Add(element);
            }

            foreach (TSource element in first)
            {
                if (set.Remove(element))
                {
                    yield return element;
                }
            }
        }
    }

    internal sealed class Set<TElement>
    {
        private readonly IEqualityComparer<TElement> _comparer;
        private int[] _buckets;
        private Slot[] _slots;
        private int _count;

        public Set(IEqualityComparer<TElement> comparer)
        {
            _comparer = comparer ?? EqualityComparer<TElement>.Default;
            _buckets = new int[7];
            _slots = new Slot[7];
        }

        // If value is not in set, add it and return true; otherwise return false
        public bool Add(TElement value)
        {
            int hashCode = InternalGetHashCode(value);
            for (int i = _buckets[hashCode % _buckets.Length] - 1; i >= 0; i = _slots[i]._next)
            {
                if (_slots[i]._hashCode == hashCode && _comparer.Equals(_slots[i]._value, value))
                {
                    return false;
                }
            }

            if (_count == _slots.Length)
            {
                Resize();
            }

            int index = _count;
            _count++;
            int bucket = hashCode % _buckets.Length;
            _slots[index]._hashCode = hashCode;
            _slots[index]._value = value;
            _slots[index]._next = _buckets[bucket] - 1;
            _buckets[bucket] = index + 1;
            return true;
        }

        // If value is in set, remove it and return true; otherwise return false
        public bool Remove(TElement value)
        {
            int hashCode = InternalGetHashCode(value);
            int bucket = hashCode % _buckets.Length;
            int last = -1;
            for (int i = _buckets[bucket] - 1; i >= 0; last = i, i = _slots[i]._next)
            {
                if (_slots[i]._hashCode == hashCode && _comparer.Equals(_slots[i]._value, value))
                {
                    if (last < 0)
                    {
                        _buckets[bucket] = _slots[i]._next + 1;
                    }
                    else
                    {
                        _slots[last]._next = _slots[i]._next;
                    }

                    _slots[i]._hashCode = -1;
                    _slots[i]._value = default(TElement);
                    _slots[i]._next = -1;
                    return true;
                }
            }

            return false;
        }

        private void Resize()
        {
            int newSize = checked((_count * 2) + 1);
            int[] newBuckets = new int[newSize];
            Slot[] newSlots = new Slot[newSize];
            Array.Copy(_slots, 0, newSlots, 0, _count);
            for (int i = 0; i < _count; i++)
            {
                int bucket = newSlots[i]._hashCode % newSize;
                newSlots[i]._next = newBuckets[bucket] - 1;
                newBuckets[bucket] = i + 1;
            }

            _buckets = newBuckets;
            _slots = newSlots;
        }

        internal TElement[] ToArray()
        {
            TElement[] array = new TElement[_count];
            for (int i = 0; i != array.Length; ++i)
            {
                array[i] = _slots[i]._value;
            }

            return array;
        }

        internal List<TElement> ToList()
        {
            int count = _count;
            List<TElement> list = new List<TElement>(count);
            for (int i = 0; i != count; ++i)
            {
                list.Add(_slots[i]._value);
            }

            return list;
        }

        internal int Count
        {
            get { return _count; }
        }

        internal int InternalGetHashCode(TElement value)
        {
            // Handle comparer implementations that throw when passed null
            return (value == null) ? 0 : _comparer.GetHashCode(value) & 0x7FFFFFFF;
        }

        internal struct Slot
        {
            internal int _hashCode;
            internal int _next;
            internal TElement _value;
        }
    }
}
```

JonHanna commented:
`Contains()` is O(n) while `Set.Remove()` is O(1) with greater constant values. As such I'd expect a `Contains()`-based approach to be appreciably faster for small input sequences but significantly slower for large input sequences. Can you show that this approach still out-performs the current implementation when `first` and `second` each contain a few hundred values?

varocarbas commented:
@JonHanna Sure. Additionally, I have realised about other scenario not being accounted properly. I will correct my submission in a short while.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,14 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
+           
+            foreach (TSource element in first)
+            {
+                if (second.Contains(element))
```
This calls `Contains` on `second` multiple times, without preparing a buffer first. There's no guarantee that the second call won't throw an exception, search an empty sequence, wait on a long-running database query before returning, search completely different values each time, access a webservice `first.Count()` times where before it was accessing it once, and so on.

JonHanna replied:
Also, the impl of `Contains` now hit will always go through `IEqualityComparer<TSource>` which you'd claimed as a problem with the original.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,14 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
-
-            return IntersectIterator(first, second, null);
+           
+            foreach (TSource element in first)
```
There's no guard against `first` containing multiple values that are in `second` and returning them each time instead of just once.
(If this passed tests, we have a gap in the testing).

varocarbas replied:
@JonHanna  This was the error I said before. Already fixed via Distinct(). I am not sure what you mean in your previous comment.

varocarbas commented:
@JonHanna  As said via comment, my mistake was forgetting about the `Distinct` peculiarities when dealing with duplicate entries, as you rightly pointed out. I fixed it via calling Distinct() for both inputs. Ah I get what you meant with your other comment. Wait one sec.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,17 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             }
+           
+            var first2 = first.Distinct();
+            var second2 = second.Distinct();
```
The current code creates a `Set<T>` which has some up-front cost, and then removes from it on the fly. This creates `1 + first.Count()` `Set<T>` so it has that up-front cost multiple times, and then loses the "knowledge" of the set so `Contains` can't be optimised. Did you re-try the benchmarks when you changed the code?

varocarbas commented:
@JonHanna PS: tested the size and isn't an issue. Let me fix this last problem and will sum all the ideas together.

JonHanna commented:
Are you running tests locally before pushing? You can run just the tests for one project from within the test directory.

varocarbas commented:
@JonHanna Sorry about a so chaotic PR. I so a so quick improvement so easily (my last PR was much more complex) than I couldn't refrain myself.

I cannot run the Linq CoreFX tests in my computer because I get an error (theoretically it was fixed months ago, this is what some of your colleagues said). Let me some minutes which I will take care of the last issues. This modification isn't too difficult and can do it without tests.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,17 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
+            {
+                if (second2.Contains(element))
+                {
+                    yield return element;
```
My previous comment still holds here, except for the possibility of accessing a web service `first.Distinct().Count()` times, etc.

varocarbas replied:
@JonHanna  I know. Please, let me finish my corrections and my tests. Just a couple of minutes more.

JonHanna replied:
Sorry, the format can lead to one asking about things already answered, especially when using a phone.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
+            IEnumerable<TSource> small = null;
 
-            return IntersectIterator(first, second, null);
+            if (first.Count() > second.Count())
```
Count can be O(n), so this could be on a similar order in itself as Intersect was before the change, depending on the source. We do have optimisations that obtain counts only when get can be reasonably fast when that offers a possible optimisation. Examine how they work.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            IEnumerable<TSource> big = null;
```
Nit: No need to set to null when every branch will set this before the first use of it.

JonHanna commented on src/System.Linq/src/System/Linq/Intersect.cs:
```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
+
+            foreach (TSource element in big)
+            {
+                if (small.Contains(element))
```
This will at best run the implementation of Distinct big.Count() times, and could still be hitting a database or web server that often, getting different results, throwing an exception, etc.

varocarbas replied:
@JonHanna You have to bear in mind that these specific Linq methods are very quick and never (not as per my knowledge) give problems. The original code adds and then removes the elements as required and through the unclear-to-me-how-optimal Set class. Just do the tests. And don't quite get what you mean with "hitting a database or web server that often".

varocarbas replied:
@JonHanna Bear in mind that I came directly to this method because of expecting a not-too-good performance (I saw very slow behaviours in quite a few situations). But I mean you are the boss here, if you don't like the suggestion I cannot do anything. Although I can do more tests tomorrow.

JonHanna replied:
I mean that `first` and/or `second` could be yielding from the results of a database or web service, and this change means instead of doing so once, it may do so multiple times.

varocarbas replied:
@JonHanna  Ah I see. But only `first.Count()` and `second.Count()` would bother on that as far as `big` and `small` are new local instances and this code yields (populates the output collection) the same number of times than the default implementation. As said, I am a bit tired now. Let me new look at it tomorrow and read all your posts.

varocarbas commented:
@JonHanna  I think that all the issues are fixed now. The code is notably bigger with various calls, but the performance is pretty much the same.

As said, I tested a higher number of elements (up to 5000) being common or different. No relevant changes in any of the alternatives. 

I cannot run the Linq tests (and some other more) in my Windows 7 computer because of the error

> The "GetTargetMachineInfo" task failed unexpectedly.

In any case, I re-apologise for having pulled this request without having properly tested it. I think that now it should be fine, but I am a bit tired and will go to bed right away. Tomorrow I will recheck it and do some proper tests.

JonHanna commented:
On the testing difficulty, have you tried doing a full git clean (build clean should suffice, but git clean removes even more) and a full build of the full repository. I often find that helps when project builds are giving me a bunch of errors I can't grok.

varocarbas commented:
@JonHanna  OK. I will try this too. But tomorrow :)

stephentoub commented:
@varocarbas, thanks for the interest in contributing.  I skimmed the conversation with @JonHanna, but a few thoughts after a quick skim of the proposed changes:

1. The test in your PR description is flawed: you're not actually iterating the results of calling Iterate or Iterate2, so it's not actually testing the real costs involved.
2. You've converted Intersect into being an iterator, i.e. it contains yield statements.  That isn't a valid change, as it means that the input enumerables are enumerated when Iterate is called rather than on the first call to MoveNext.
3. As @JonHanna has called out, this changes the number of times the inputs are enumerated, which is not a valid change.  Each call to .Count() is going to enumerate the source, at least for some source types (e.g. ones that don't implement ```ICollection<T>```), as will each iteration of the result of Distinct, each call to Contains, etc.
4. When I fix issue (1) above, my tests show your new version performing an order of magnitude worse than the current implementation, and even more so when the inputs aren't ```ICollection<T>```, e.g. when the inputs are themselves iterators.

Again, appreciate the interest, but both the product changes and the validation of those changes need work before this can be further considered.

JonHanna commented:
There's a flaw in your benchmark:

```C#
            sw.Start();

            while (count < maxCount)
            {
                count++;
                Intersect(first, second);
                //Intersect2(first, second); 
            }
            sw.Stop();
```

This doesn't test the actual cost of iterating through the results, just the initial set-up of the iterator which tends to be much smaller than the iteration cost.
When benchmarking iteration I tend to use an empty `foreach()` or a method like:

```C#
static ConsumeAll<T>(IEnumerable<T> source)
{
  using(en = source.GetEnumerator())
    while(en.MoveNext()){}
}
```
Which is a pretty light consumption, and therefore not measuring too much of something unrelated, though sometimes it's good to hit `Current` too.

varocarbas commented:
@JonHanna @stephentoub 

Firstly, I re-(re-)apologise for this a bit impulsive PR and the initial faulty versions of the code (even the last version doesn't account for all the scenarios, but I will not improve it because this PR should be closed). I thought that the improvement was so clear, was pretty sure about `Intersect` being slow (quite a few problems on this front in the past) and my last PR took so long that I was too impatient.

Regarding the CoreFX tests, note that I initially cloned the repository via Visual Studio 2015 to a blank folder and cleaned/built everything. Now, I tried via git app by also doing a fresh copy and cleaning everything. Still same error.

Regarding my suggestion and your last comments, you are right. My approach (or variations of it) isn't iteration/`yield` friendly, what this is precisely about. The original algorithm adds and removes elements to a third collection what seems unnecessarily time-consuming (at least, when you can rely on efficient alternatives like `Contains` accounting for the kind of comparison you want), but apparently this isn't that bad within an iterator (as proven by this PR, my experience on this specific micro-field isn't too relevant).

It is kind of curious though (I personally have learned something, not sure about you) the so relevant differences between getting the collection (what my original tests were measuring, where a properly-working version of my proposal is much quicker than the original code) and iterating directly through it (the testing approach suggested in your last posts, where any version of my proposal is notably slower than the original code). Shouldn't this reality be accounted for and an additional method be included? One version for iterations and another (notably quicker) version for getting the collection?

In any case, I guess that this issue should be closed as far as the proposed improvement isn't such (at least, not in all the scenarios).

varocarbas commented:
@JonHanna @stephentoub
(I removed my previous comment. Include the worthy bits here)

Firstly, I re-(re-)apologise for this a bit impulsive PR and the initial faulty versions of the code (even the last version doesn't account for all the scenarios, but I will not improve it because this PR should be closed). I thought that the improvement was so clear, was pretty sure about Intersect being slow (quite a few problems on this front in the past) and my last PR took so long that I was too impatient.

Regarding the CoreFX tests, note that I initially cloned the repository via Visual Studio 2015 to a blank folder and cleaned/built everything. Now, I tried via git app by also doing a fresh copy and cleaning everything. Still same error.

You are right regarding notably different results when testing the iterations (as suggested by JonHanna via an empty `foreach` or a code on the proposed lines). In any case, this is certainly different than what my original code was testing, although this doesn't mean that it flawed. Both realities (getting a collection or iterating through it) are two different conditions which have to be adequately tested. Any improvement needs to perform better than the original code in both of them. At a first sight, it seemed that my suggestion performed notably worse in these iterating scenarios, but after a closer look (and improving yesterday's not-that-good code), it might also be better here. For small number of elements both approaches are really similar (only in this iterating alternative), but for big number of elements my approach seems notably better.

I will do a proper set of tests and come back to you once my ideas are completely clear.

stephentoub commented:
> as rightly pointed out via comments, the original test didn't account for the iterator peculiarities 

It's also only testing ```List<T>```.  You need to try it with sources that don't implement ```ICollection<T>``` to provide an efficient Contains implementation; the new version will fare notably worse with such sources.

varocarbas commented:
@JonHanna @stephentoub
After doing more tests, I am afraid that my first impression this morning was right: for the iterator scenario (e.g., being called within a `foreach` loop), the existing version is some times better; although my suggestion always performs better for what I called `TestReturnedCollection` (i.e., my original scenario, just getting the output without iterations). Note that I have updated my testing code and now accounts for these two eventualities. In the `TestIterator` scenario both approaches are very similar in many situations, mainly for a low number of items/iterations and when most of items are ignored (e.g., because of being identical). But there are quite a few situations (like the one being showed right now by the code) where the unmodified version is much better. I guess that this is enough to conclude that better sticking to what is known and delete this PR.

In any case, this issue of having two different situations (getting the collection right away or iterating through it) with notably different performances is something I wasn't aware of. Perhaps it might be good to add new methods performing notably better, expected to not be iterated? Anyway, I guess that this isn't part of the current PR, whose point was suggesting a quicker alternative which isn't such (at least, not under all the possible scenarios).

varocarbas commented:
@stephentoub My approach (here or in general) is finding for-sure-performing better bits and then keep extending the operating range. I used lists as a first step. But your and JonHanna's suggestions of also considering the performance while iterating have been proven too much. The existing approach has an overall better performance on these scenarios and that's why my suggestion cannot be considered better on this front. Proposing an improvement which might provoke notably worse performances in certain cases doesn't seem right. That's why I think that you should close this ticket, although I see some relevant issues here which perhaps you should think about.

stephentoub commented:
> That's why I think that you should close this ticket

Ok. Thanks.

varocarbas commented:
Thanks to both of you.

JonHanna commented:
I've certainly also had ideas I submitted because it improved one benchmark, that then turned out not to do so well when considered across a wider range of cases in the past. And by "in the past" I mean this week 😄 

@varocarbas did doing a full clean and build of the repository fix your difficulty running tests locally? If not it'd be worth opening an issue about, because it really stymies trying to contribute when that's not working for you.

varocarbas commented:
@JonHanna No. I cannot run `Linq` tests and quite a few other ones. Note that I don't keep a local CoreFX repository, but download the latest one every time I want to work on this. During the last months, I have always been getting errors in some of the tests. Around 5 months ago, I had Windows 10 installed and didn't see any of this (now I am back to Windows 7). Shall I open a normal CoreFX issue?

JonHanna commented:
Do. It's something that should work, so either you're doing something wrong or something is broken, and either way we should get that fixed.

varocarbas commented:
I will do it right away. Thanks.
